src/corelibs/U2Formats/src/BAMUtils.cpp

```diff
@@ -606,27 +631,40 @@ static int bamMergeCore(const QString& outFileName, const QList<QString>& filesT
     return 0;
 }
 
-GUrl BAMUtils::mergeBam(const QStringList& bamUrls, const QString& mergedBamTargetUrl, U2OpStatus& os) {
-    coreLog.details(tr(R"(Merging BAM files: "%1". Resulting merged file is: "%2")")
-                        .arg(QString(bamUrls.join(",")))
-                        .arg(QString(mergedBamTargetUrl)));
+static constexpr int MAX_FILES_OPENED = 100;
+static int recursiveBamMergeCore(const QString& outFileName, const QList<QString>& filesToMerge) {
+    int size = filesToMerge.size();
+    CHECK(size != 0, -1);
 
-    int rc = bamMergeCore(mergedBamTargetUrl, bamUrls);
-    CHECK_EXT(rc >= 0, os.setError(tr("Failed to merge BAM files: %1 into %2").arg(bamUrls.join(",")).arg(mergedBamTargetUrl)), {});
-    return mergedBamTargetUrl;
-}
+    int remainder = size % MAX_FILES_OPENED;
+    int iterationNumber = size / MAX_FILES_OPENED + (remainder != 0 ? 1 : 0);
+    if (iterationNumber == 1) {
+        return localBamMergeCore(outFileName, filesToMerge);
+    }
 
-void* BAMUtils::loadIndex(const QString& filePath) {
-    // See bam_index_load_local.
-    QString mode = "rb";
-    NP<FILE> file = BAMUtils::openFile(filePath + ".bai", mode);
-    if (file == nullptr && filePath.endsWith("bam")) {
-        file = BAMUtils::openFile(filePath.chopped(4) + ".bai", mode);
+    QStringList newOutFileNameList;
+    for (int i = 0; i < iterationNumber; i++) {
+        int midLength = MAX_FILES_OPENED;
+        if (i == iterationNumber - 1) {
+            midLength = remainder;
+        }
+        QList<QString> newFilesToMerge = filesToMerge.mid(i * MAX_FILES_OPENED, midLength);
+        auto firstFileToMergeName = newFilesToMerge.first();
+        // Remove ".bam" from the end
+        QString newOutFileBaseName = firstFileToMergeName.left(firstFileToMergeName.size() - 4);
+        QString newOutFileName = newOutFileBaseName + "_" + QString::number(i) + ".bam";
+        newOutFileNameList << newOutFileName;
+        int res = recursiveBamMergeCore(newOutFileName, newFilesToMerge);
+        CHECK(res >= 0, res);
     }
-    CHECK(file != nullptr, nullptr);
-    bam_index_t* idx = bam_index_load_core(file);
-    closeFileIfOpen(file);
-    return idx;
+
+    return recursiveBamMergeCore(outFileName, newOutFileNameList);
```

Comment:
Could you please document why we need the recursion here? Can it be done with no recursion if we use U2Region:split?

Reply:
Thanks for using U2Region::split it made the code simpler and now I can see the logic of the method clearer

Question:
Once you did `auto mergeSplit = U2Region::split({ 0, size }, MAX_FILES_OPENED);`
the `newFilesToMerge` size is < MAX_FILES_OPENED and you can call `localBamMergeCore` instead of `recursiveBamMergeCore` right?

Reply:
Yes, you are right, it will work. I added the recursion here because of the following problem: if we have 1'000 files to merge, we split them to 10 bunches 100 files in each bunch. In the result, we have 10 files (each file is a result of merging of each bunch), then we merge these 10 files and everything is OK, it works.

But, if we have 100'000 files to merge, we split them into 1'000 bunches to merge (100 files in each bunch). So, in the result, we have 1'000 files (one file is the merge result from each bunch) and it's still to many to merge in one time, we need to split them again. This recursive algorithm prevent this scenario.

Actually, it may be redundantly, because the file which will be split to 100'000 should be **extremely** large and I've never met such a big data, but, just several weeks ago we've never met SAM file about 200Gb (which were sent us by Roman, who found this bug).

Reply:
The line I mentioned
`int res = recursiveBamMergeCore(newOutFileName, newFilesToMerge)`
there is no need in recursion call because `newFilesToMerge.size()` is always < MAX_FILES_OPENED, right?

BTW could you please add a comment which problem is being solved by the function. This functionality is not used by the original samtools right?

Reply:
OK

Reply:
> BTW could you please add a comment which problem is being solved by the function. This functionality is not used by the original samtools right?

What function do you mean - `bamMergeCore` or `recursiveBamMergeCore`? Anyway, this functionality is not used by samtools, this functionality **uses** samtools.

Reply:
I know what `bamMergeCore` does because I created it. I mean `recursiveBamMergeCore`. Could you please document the problem it solves and why the original samtools has no such problem?

I have 3 more questions about this function: 
1. Who cleans up the tmp files you created during recursion?  Could you please add a check to the test that no tmp files are left after the merge.
2. Why do you think that the first file location in a bucket is a good place for tmp files? 
3. About the name of the tmp merge file: please add a check that this name does not overlap with any of existing files - this name must be a unique name.

Reply:
> 1. Who cleans up the tmp files you created during recursion? Could you please add a check to the test that no tmp files are left after the merge.
 No one does because no one has been doing it before. All intermediate files, which are created during importing SAM/BAM to UGENEDB, are saved and I remember this behavior has had a place during the whole time I remember. I can change this behavior if you want, but please - not in this issue.

> 2. Why do you think that the first file location in a bucket is a good place for tmp files?
Don't get the point. What is "bucket"? What first file? What location?

> 3. About the name of the tmp merge file: please add a check that this name does not overlap with any of existing files - this name must be a unique name.

Let me do it in another issue, please. Current files are overwritten and this behavior has had a place during years.

Reply:
Moving discussion into a separate thread: 1 per issue.